### ISSUE TITLE
[BUG] Warning events are being spammed by Longhorn - CRD

### ISSUE THREAD

**Issue by larivierec** [longhorn/longhorn#7290]
## Describe the bug (🐛 if you encounter this issue)

Warning events are being spammed by Longhorn because the Longhorn Node CRD doesn't seem to recognize a new message in the latest k3s release.

"Unknown condition true of Kubernetes node <name>: condition is of the EtcdIsVoter reason is, MemberNotLearner, message is Node is a voting member of etcd cluster."
## To Reproduce

- Install latest k3s 1.28.4-k3s2 in cluster mode --cluster-init (for etcd) 

- Install Longhorn chart

Nodes that are ready will broadcast a new condition not known by the Node CRD v1beta2 and spam warning events in longhorn-system namespace.

## Expected behavior

No events or logs should be present.

## Environment
Longhorn: 1.5.3
K3s: 1.28.4-k3s2

3 master nodes
5 workers

Ubuntu 22.04 baremetal.

**Comment by m-ildefons** [comment]
This warning is caused by K3s setting a new node condition. This was introduced with c5cd7b3d6543ef782c84651b5c46e904ca83828b, after an [issue](https://github.com/k3s-io/k3s/issues/8033) and [adr](https://github.com/k3s-io/k3s/blob/6ba6c1b65f9483a5eb3657206ca58c9a7464ad9d/docs/adrs/status-for-etcd-node.md?plain=1#L2)
Various K8s implementations or cloud providers are allowed to set their own node conditions and the set provided in the K8s API is just a minimal set guaranteed to be there: https://pkg.go.dev/k8s.io/api/core/v1#NodeConditionType

There is nothing to worry about with this particular node condition as it only relays information about which node's etcd instance takes what kind of responsibility, so adding and removing nodes from a k3s cluster taking this information into account doesn't jeopardize the quorum of the etcd cluster.

Longhorn exposes all unknown (to longhorn) node conditions as K8s events. This is a mechanism that only has informational value and is used as such in Longhorn as well, as it only serves to relay information to the UI. The warnings are harmless in this case.

An argument can be made to ignore all unknown node conditions by default, or alternatively just ignore this particular one. What do you think @innobead , should Longhorn assume that an unknown node condition is worthy of being a warning, just as it is now?
It should be noted here that events are only emitted for node conditions that are "true", but not for ones that are "false". However there is no convention or guidance if "true" or "false" should indicate an error condition or healthiness.
I think the best course of action would be to just ignore non-standard node conditions except when we can derive value from them.

**Comment by larivierec** [comment]
Understood.

Yeah, I saw the commit in the K3s repository yesterday and noticed it's a new addition from 3 weeks ago.

The reason for this issue is mainly because we should have a little bit of control to what should be logged in especially when it's harmless.

```bash
(⎈|k3s:default)➜  home-cluster git:(main) ✗ k get events -n longhorn-system
LAST SEEN   TYPE      REASON                     OBJECT           MESSAGE
4m9s        Warning   UnknownNodeConditionTrue   node/fluffy      Unknown condition true of kubernetes node fluffy: condition type is EtcdIsVoter, reason is MemberNotLearner, message is Node is a voting member of the etcd cluster
4m9s        Warning   UnknownNodeConditionTrue   node/fluffy      Unknown condition true of kubernetes node fluffy: condition type is EtcdIsVoter, reason is MemberNotLearner, message is Node is a voting member of the etcd cluster
4m9s        Warning   UnknownNodeConditionTrue   node/fluffy      Unknown condition true of kubernetes node fluffy: condition type is EtcdIsVoter, reason is MemberNotLearner, message is Node is a voting member of the etcd cluster
4m12s       Warning   UnknownNodeConditionTrue   node/fluffy      Unknown condition true of kubernetes node fluffy: condition type is EtcdIsVoter, reason is MemberNotLearner, message is Node is a voting member of the etcd cluster
4m9s        Warning   UnknownNodeConditionTrue   node/fluffy      Unknown condition true of kubernetes node fluffy: condition type is EtcdIsVoter, reason is MemberNotLearner, message is Node is a voting member of the etcd cluster
4m9s        Warning   UnknownNodeConditionTrue   node/fluffy      Unknown condition true of kubernetes node fluffy: condition type is EtcdIsVoter, reason is MemberNotLearner, message is Node is a voting member of the etcd cluster
4m9s        Warning   UnknownNodeConditionTrue   node/fluffy      Unknown condition true of kubernetes node fluffy: condition type is EtcdIsVoter, reason is MemberNotLearner, message is Node is a voting member of the etcd cluster
2m38s       Warning   UnknownNodeConditionTrue   node/fluffy      Unknown condition true of kubernetes node fluffy: condition type is EtcdIsVoter, reason is MemberNotLearner, message is Node is a voting member of the etcd cluster
3m15s       Warning   UnknownNodeConditionTrue   node/frenzy      Unknown condition true of kubernetes node frenzy: condition type is EtcdIsVoter, reason is MemberNotLearner, message is Node is a voting member of the etcd cluster
3m15s       Warning   UnknownNodeConditionTrue   node/frenzy      Unknown condition true of kubernetes node frenzy: condition type is EtcdIsVoter, reason is MemberNotLearner, message is Node is a voting member of the etcd cluster
3m15s       Warning   UnknownNodeConditionTrue   node/frenzy      Unknown condition true of kubernetes node frenzy: condition type is EtcdIsVoter, reason is MemberNotLearner, message is Node is a voting member of the etcd cluster
3m15s       Warning   UnknownNodeConditionTrue   node/frenzy      Unknown condition true of kubernetes node frenzy: condition type is EtcdIsVoter, reason is MemberNotLearner, message is Node is a voting member of the etcd cluster
3m15s       Warning   UnknownNodeConditionTrue   node/frenzy      Unknown condition true of kubernetes node frenzy: condition type is EtcdIsVoter, reason is MemberNotLearner, message is Node is a voting member of the etcd cluster
3m15s       Warning   UnknownNodeConditionTrue   node/frenzy      Unknown condition true of kubernetes node frenzy: condition type is EtcdIsVoter, reason is MemberNotLearner, message is Node is a voting member of the etcd cluster
3m15s       Warning   UnknownNodeConditionTrue   node/frenzy      Unknown condition true of kubernetes node frenzy: condition type is EtcdIsVoter, reason is MemberNotLearner, message is Node is a voting member of the etcd cluster
2m38s       Warning   UnknownNodeConditionTrue   node/frenzy      Unknown condition true of kubernetes node frenzy: condition type is EtcdIsVoter, reason is MemberNotLearner, message is Node is a voting member of the etcd cluster
2m9s        Warning   UnknownNodeConditionTrue   node/whirlwind   Unknown condition true of kubernetes node whirlwind: condition type is EtcdIsVoter, reason is MemberNotLearner, message is Node is a voting member of the etcd cluster
2m9s        Warning   UnknownNodeConditionTrue   node/whirlwind   Unknown condition true of kubernetes node whirlwind: condition type is EtcdIsVoter, reason is MemberNotLearner, message is Node is a voting member of the etcd cluster
2m9s        Warning   UnknownNodeConditionTrue   node/whirlwind   Unknown condition true of kubernetes node whirlwind: condition type is EtcdIsVoter, reason is MemberNotLearner, message is Node is a voting member of the etcd cluster
2m9s        Warning   UnknownNodeConditionTrue   node/whirlwind   Unknown condition true of kubernetes node whirlwind: condition type is EtcdIsVoter, reason is MemberNotLearner, message is Node is a voting member of the etcd cluster
2m9s        Warning   UnknownNodeConditionTrue   node/whirlwind   Unknown condition true of kubernetes node whirlwind: condition type is EtcdIsVoter, reason is MemberNotLearner, message is Node is a voting member of the etcd cluster
2m9s        Warning   UnknownNodeConditionTrue   node/whirlwind   Unknown condition true of kubernetes node whirlwind: condition type is EtcdIsVoter, reason is MemberNotLearner, message is Node is a voting member of the etcd cluster
2m9s        Warning   UnknownNodeConditionTrue   node/whirlwind   Unknown condition true of kubernetes node whirlwind: condition type is EtcdIsVoter, reason is MemberNotLearner, message is Node is a voting member of the etcd cluster
3m7s        Warning   UnknownNodeConditionTrue   node/whirlwind   Unknown condition true of kubernetes node whirlwind: condition type is EtcdIsVoter, reason is MemberNotLearner, message is Node is a voting member of the etcd cluster
```

Right now, for each of my master nodes, it seems to be emitting 7 messages / 60 seconds.
Could `unknown` conditions be ignored through a chart option to the longhorn application perhaps?

**Comment by aumer-amr** [comment]
I second this, after upgrading it's quite spamming the events. An ignore on these unknown conditions by default, or giving the option would be highly appreciated.

**Comment by LilDrunkenSmurf** [comment]
FYI for context, I don't get these messages, however my control planes are tainted as `NoSchedule`, so Longhorn never runs on my control plane nodes.

**Comment by brandond** [comment]
This comes from here: https://github.com/longhorn/longhorn-manager/blob/268b1d63b9570e750bc13e011f475a17e6878041/controller/node_controller.go#L455-L458

Apparently longhorn will barf if any unknown condition is set to true.

**Comment by PhanLe1010** [comment]
> An argument can be made to ignore all unknown node conditions by default, or alternatively just ignore this particular one. What do you think @innobead , should Longhorn assume that an unknown node condition is worthy of being a warning, just as it is now?
It should be noted here that events are only emitted for node conditions that are "true", but not for ones that are "false". However there is no convention or guidance if "true" or "false" should indicate an error condition or healthiness.
I think the best course of action would be to just ignore non-standard node conditions except when we can derive value from them.

Ignoring unknown condition looks reasonable to me. The other approach is at least we should not repeatedly spam the API server by only emitting the unknown condition once?

**Comment by brandond** [comment]
My personal preference would be to have LH ignore any conditions that it doesn't specifically derive information from. If users want to be alerted about specific conditions on their node, there are tools other than LH that will handle this.

**Comment by innobead** [comment]
> My personal preference would be to have LH ignore any conditions that it doesn't specifically derive information from. If users want to be alerted about specific conditions on their nodes, there are tools other than LH that will handle this.

Agreed. Should just do it explicitly by watching what conditions we care about. @m-ildefons Please continue working on it. Thanks.

**Comment by ijorjadze** [comment]
i find those spummy events today , so fix will be on v1.6.0 version ?

**Comment by larivierec** [comment]
any updates on this?

**Comment by innobead** [comment]
This will be fixed in the upcoming 1.6.0 and backported to 1.5.4 & 1.4.5.

**Comment by roger-ryao** [comment]
**Verified on master-head 20231225**

- longhorn master-head https://github.com/longhorn/longhorn/commit/fe9dc3c5530f58b80085686718e351a8e4634229
- longhorn-manager master-head https://github.com/longhorn/longhorn-manager/commit/859d43893b67a3cc86d83938f33d9c38673e3636

**The test steps**

https://github.com/longhorn/longhorn/issues/7290#issue-2031804175

1. Install k3s Server and Initialize the Cluster
**NOTE: For this test case, node taints `node-role.kubernetes.io/master=true:NoExecute` & `node-role.kubernetes.io/master=true:NoSchedule` on the control plane node have NOT been added.**
```
curl -sfL https://get.k3s.io | K3S_TOKEN=SECRET INSTALL_K3S_EXEC="server --cluster-init" sh -
```
2. Check Control Node's Conditions
```
# Replace CONTROL_NODE_NAME with the actual name of your control node
kubectl get node $CONTROL_NODE_NAME -o=jsonpath='Node Name: {.metadata.name}{"\n"}Conditions:{"\n"}{range .status.conditions[*]}- Type: {.type}{"\n"}  Status: {.status}{"\n"}  LastHeartbeatTime: {.lastHeartbeatTime}{"\n"}  LastTransitionTime: {.lastTransitionTime}{"\n"}  Reason: {.reason}{"\n"}  Message: {.message}{"\n\n"}{end}'
```
3. Install Longhorn
4. Check events in the longhorn-system namespace

```
kubectl -n longhorn-system get events
```

**Result Passed**

1. Installed Longhorn `v1.5.3` on `k3s cluster` and observed warnings in longhorn-system namespace:

```console
65s         Warning   UnknownNodeConditionTrue   node/ip-10-0-1-57   Unknown condition true of kubernetes node ip-10-0-1-57: condition type is EtcdIsVoter, reason is MemberNotLearner, message is Node is a voting member of the etcd cluster
39s         Warning   UnknownNodeConditionTrue   node/ip-10-0-1-57   Unknown condition true of kubernetes node ip-10-0-1-57: condition type is EtcdIsVoter, reason is MemberNotLearner, message is Node is a voting member of the etcd cluster
38s         Warning   UnknownNodeConditionTrue   node/ip-10-0-1-57   Unknown condition true of kubernetes node ip-10-0-1-57: condition type is EtcdIsVoter, reason is MemberNotLearner, message is Node is a voting member of the etcd cluster
24s         Warning   UnknownNodeConditionTrue   node/ip-10-0-1-57   Unknown condition true of kubernetes node ip-10-0-1-57: condition type is EtcdIsVoter, reason is MemberNotLearner, message is Node is a voting member of the etcd cluster
```

2. Inspected node conditions and found new condition `EtcdIsVoter`:

```console
- Type: EtcdIsVoter
  Status: True
  LastHeartbeatTime: 2023-12-25T14:16:26Z
  LastTransitionTime: 2023-12-25T14:16:26Z
  Reason: MemberNotLearner
  Message: Node is a voting member of the etcd cluster
```

3. Checked control plane node events in Longhorn `master-head` and did not observe `EtcdIsVoter` warnings:

**Comment by Ashkaan** [comment]
I tried Longhorn and it didn't work for my need.  I removed it from my cluster, but I still have the warnings (in Lens).  How do I clear those warnings?

**Comment by Sierra1011** [comment]
Any idea when we can expect 1.4.5? I accidentally encountered this bug while on 1.4 right before Christmas (idle hands make for poor decisions :grin: ) and this has crippled my homelab.

**Comment by Ashkaan** [comment]
Have you figured out how to clear the warnings?  There must be some way to edit ETCD (or wherever that's stored).

**Comment by omidraha** [comment]
I also received this `warning` when I installed `longhorn` on the specific node.
https://github.com/longhorn/longhorn/issues/7407#issuecomment-1881838982

**Comment by PhanLe1010** [comment]
> I tried Longhorn and it didn't work for my need. I removed it from my cluster, but I still have the warnings (in Lens). How do I clear those warnings?

Hi @Ashkaan What warning are you still seeing? Have you follow the instruction to uninstall Longhorn https://longhorn.io/docs/1.5.3/deploy/uninstall/ ?

**Comment by PhanLe1010** [comment]
> 
Any idea when we can expect 1.4.5? I accidentally encountered this bug while on 1.4 right before Christmas (idle hands make for poor decisions 😁 ) and this has crippled my homelab.

Hi @Sierra1011 Do you mean the cluster is frozen?

**Comment by brandond** [comment]
Until the new release of LH is available, the only way to stop getting the events is to downgrade to an older release of K3s/RKE2.

**Comment by Ashkaan** [comment]
> > I tried Longhorn and it didn't work for my need. I removed it from my cluster, but I still have the warnings (in Lens). How do I clear those warnings?
> 
> Hi @Ashkaan What warning are you still seeing after uninstall Longhorn? Have you followed the instructions to uninstall Longhorn https://longhorn.io/docs/1.5.3/deploy/uninstall/ ?

Yes, I followed this:
```helm uninstall longhorn -n longhorn-system```

I don't have anything longhorn installed anywhere (so far as I can tell) and I get this:

<img width="2175" alt="image" src="https://github.com/longhorn/longhorn/assets/5667978/9164823b-5fe5-469d-a103-f1bd2df08705">

**Comment by brandond** [comment]
You get that where? Those events are 23 days old. Normally events don't hang around that long.

**Comment by Ashkaan** [comment]
Correct.  These have been lingering and driving our team crazy.  We can see them in Lens.  It happened as soon as we tested longhorn for a project.

**Comment by brandond** [comment]
Have you tried deleting the events?

Has someone customized your cluster configuration to retain events for longer than usual? The default event ttl is a couple hours, if I remember correctly.

**Comment by Ashkaan** [comment]
Sadly, I haven't yet learned how to delete events or customize the TTL.  If you have any docs or advice, I'd really appreciate it.

Also, all other warnings go away within a couple of hours.  These always stay.

**Comment by Sierra1011** [comment]
> > Any idea when we can expect 1.4.5? I accidentally encountered this bug while on 1.4 right before Christmas (idle hands make for poor decisions 😁 ) and this has crippled my homelab.
> 
> Hi @Sierra1011 Do you mean the cluster is frozen because of the warning events?

TL;DR I am keen to know in what sort of timeframe a backport will become available.

Even after downgrading and reverting K3s versions etc, I continued getting these alerts preventing volumes being mounted. I tried removing taints by hand and deleting all the events, but no cigar. So I can't upgrade K3s version above a certain point without stopping volumes working.

It actually started/aligned with the start of a whole [saga of storage issues](https://www.reddit.com/r/openSUSE/comments/191y9xg/nfs_and_microos_mount_oddities_or_how_i_slowly/) for me that took a lot of fixing, and while I'm not certain exactly of the root cause, it's a lot of coincidence.

I'm not trying to be petulant, apologies if I'm coming across that way; I'm reluctant to do a knee-jerk upgrade from V1 to V2 when there's only one supported release that won't break with any vaguely recent version of K3s. 
TIA

**Comment by brandond** [comment]
> I continued getting these alerts preventing volumes being mounted. 

These events are purely informational and do not prevent volumes from being mounted. If you have having problems with your LH volumes, look further for the actual root cause. This is not the source of your issue.

**Comment by jLemmings** [comment]
Is there any workaround for this issue? I'm also getting flooded with multiple 100'00 errors...
Kubernetes: v1.26.11 rke2r1
Longhorn: v1.5.3

**Comment by Sierra1011** [comment]
I left it another few weeks and tried the upgrade again, prompted by Flux 2.2 requiring K8s API version >=1.26.

Still on LH 1.4.4. As soon as I upgraded to 1.26(.13.k3s2), it will not provision new PVCs. Seems as though this is likely to be unrelated so will investigate further but consider me out for now. 
Thanks!

**Comment by cs-shadowbq** [comment]
Rancher UI + Longhorn is still only listed as  103.2.1+up1.5.3 ... https://github.com/rancher/charts/tree/release-v2.9/charts/longhorn  .. still waiting for 1.5.4 or 1.6.0 to drop

**Comment by brandond** [comment]
That is a Rancher issue. Not Longhorn.

**Comment by Delta1977** [comment]
No, the root case belongs to longhorn and is solved in longhorn     1.5.4 and 1.6.0. longhorn 1.5.4 is shot before release in rancher AppStore

**Comment by brandond** [comment]
LH doesn't report anything. It just creates Kubernetes events, same as many other components. Rancher shows events in the UI. This is all off-topic; the LH chart for Rancher will be updated when the Rancher team gets to it.

**Comment by Delta1977** [comment]
its already released and LH 1.5.4 solved the log spamming !!!
https://github.com/rancher/charts/commit/159d5114c62dd567723fc749318cad71f86906e7

**Comment by vember31** [comment]
@Ashkaan  did you ever figure out how to remove those 'Node is a voting member of the etcd cluster' warning messages from Lens? I know they aren't specifically events,  they're node conditions...the events themselves are no longer being spammed by longhorn anymore thanks to this fix here.  But I also haven't figured out how to stop them from showing on the Cluster page in Lens.

Maybe it's Lens that needs to be updated?

**Comment by Ashkaan** [comment]
No, I deleted the cluster over it.  Unreal.  I’ll never try Longhorn again.
